crates/curp/src/tracker.rs

```diff
@@ -0,0 +1,351 @@
+///
+/// `010000000101_110111011111_000001000000000`
+///  ^        ^                ^    ^
+///  |        |________________|____|
+///  |________|         len    |____|
+///     head                    tail
+#[derive(Debug, Clone)]
+struct BitVecQueue {
+    /// Bits store
+    store: VecDeque<usize>,
+    /// Head length indicator
+    head: usize,
+    /// Tail length indicator
+    tail: usize,
+}
+
+impl Default for BitVecQueue {
+    #[inline]
+    fn default() -> Self {
+        Self::with_capacity(DEFAULT_BIT_VEC_QUEUE_CAP)
+    }
+}
+
+#[allow(clippy::arithmetic_side_effects, clippy::unwrap_used)] // They are checked
+impl BitVecQueue {
+    /// New with a capacity (in bits)
+    fn with_capacity(cap_bits: usize) -> Self {
+        Self {
+            store: VecDeque::with_capacity(cap_bits / USIZE_BITS),
+            head: 0,
+            tail: 0,
+        }
+    }
+
+    /// Get the length
+    fn len(&self) -> usize {
+        if self.store.is_empty() {
+            return 0;
+        }
+        (self.store.len() - 1) * USIZE_BITS + self.tail - self.head + 1
+    }
+
+    /// Get the bit value
+    fn get(&self, idx: usize) -> Option<bool> {
+        let idx = self.head + idx;
+        let index = idx / USIZE_BITS;
+        let slot = idx % USIZE_BITS;
+        if index == self.store.len() - 1 && slot > self.tail {
+            return None;
+        }
+        let bits = self.store.get(index)?;
+        Some(bits & (1 << (USIZE_BITS - 1 - slot)) != 0)
+    }
+
+    /// Set the bit value
+    fn set(&mut self, idx: usize, v: bool) {
+        let idx = self.head + idx;
+        let index = idx / USIZE_BITS;
+        let slot = idx % USIZE_BITS;
+        if index == self.store.len() - 1 && slot > self.tail {
+            return;
+        }
+        let Some(bits) = self.store.get_mut(index) else {
+            return;
+        };
+        if v {
+            *bits |= 1 << (USIZE_BITS - 1 - slot);
+        } else {
+            *bits &= !(1 << (USIZE_BITS - 1 - slot));
+        }
+    }
+
+    /// Push a value in the back
+    fn push(&mut self, v: bool) {
+        if self.tail == (USIZE_BITS - 1) || self.store.is_empty() {
+            self.tail = 0;
+            if v {
+                self.store.push_back(1 << (USIZE_BITS - 1));
+            } else {
+                self.store.push_back(0);
+            }
+            return;
+        }
+        self.tail += 1;
+        if v {
+            let last = self.store.back_mut().unwrap();
+            *last |= 1 << (USIZE_BITS - 1 - self.tail);
+        }
+    }
+
+    /// Peek the front of the bit queue
+    fn front(&self) -> Option<bool> {
+        let front = self.store.front()?;
+        Some(front & (1 << (USIZE_BITS - 1 - self.head)) != 0)
+    }
+
+    /// Peek the front usize in store and take a function to check it,
+    /// Used to batch accelerate check.
+    fn front_by(&self, f: impl Fn(usize) -> bool) -> Option<bool> {
+        self.store.front().copied().map(f)
+    }
```

Comment:
I think it would be better to name it `map_front`

Reply:
done